website/components/Events.tsx

```diff
@@ -0,0 +1,64 @@
+import { Head } from "$fresh/runtime.ts";
+import { AnalyticsEvent } from "../../commerce/types.ts";
+import { scriptAsDataURI } from "../../utils/dataURI.ts";
+
+type EventHandler = (event?: AnalyticsEvent) => void | Promise<void>;
+
+type EventsAPI = {
+  dispatch: (event: unknown) => void;
+  subscribe: (
+    handler: EventHandler,
+    options?: AddEventListenerOptions | boolean,
+  ) => () => void;
+};
+
+declare global {
+  interface Window {
+    DECO_ANALYTICS: Record<
+      string,
+      // deno-lint-ignore no-explicit-any
+      (action: string, eventType: string, props?: any) => void
+    >;
+    DECO_SITES_STD: { sendAnalyticsEvent: (event: unknown) => void };
+    DECO: { events: EventsAPI };
+  }
+}
+
+/**
+ * This function handles all ecommerce analytics events.
+ * Add another ecommerce analytics modules here.
+ */
+const snippet = () => {
+  const target = new EventTarget();
+
+  const dispatch: EventsAPI["dispatch"] = (event: unknown) => {
+    target.dispatchEvent(new CustomEvent("analytics", { detail: event }));
+  };
+
+  const subscribe: EventsAPI["subscribe"] = (handler, opts) => {
+    // deno-lint-ignore no-explicit-any
+    const cb = ({ detail }: any) => handler(detail);
+
+    target.addEventListener("analytics", cb, opts);
+
+    return () => {
+      target.removeEventListener("analytics", cb, opts);
+    };
+  };
+
+  window.DECO_SITES_STD = { sendAnalyticsEvent: dispatch };
+  window.DECO = {
+    ...window.DECO,
+    events: { dispatch, subscribe },
+  };
+};
+
+function Events() {
+  return (
+    <Head>
+      <script defer id="deco-events" src={scriptAsDataURI(snippet)} />
```

Comment:
Why is needed to use URI? Is it to leverage from `script + defer` ?

Reply:
Yes, apparently, defer does not work with inline scripts

Reply:
Yes, but it doesn't guarantee the script execution order. Defer makes sure the scripts execute in the right order

Reply:
https://gist.github.com/jakub-g/385ee6b41085303a53ad92c7c8afd7a6
According to this gist, the type inline script with type module implies `defer`.

Reply:
I mean, 
```
<script src="script-1"  defer />
<script src="script-2"  defer />
```

I know that script-1 will run before script-2. With type="module" I do not have that certainty